### DISCUSSION TITLE
One instance for multiple sites?

### DISCUSSION THREAD

**Discussion by mipimipi** [umputun/remark42#950] [Q&A]
Is it possible that I use one instance of remark42 for multiple sites (i.e. a web server is serving multiple sites and I want to use remark42 for all of them)? At least the documentation in the README suggests that (SITE parameter can have multiple values). But I did not find any documentation how to achieve this.
Otherwise, I assume that one must run one container instance per site. Correct?
Thanks, mipi

**Comment by umputun** [comment]
you can pass --site multiple times, i.e. `--site=foo --site=bar` or in env, comma-separated like `--SITE=foo,bar` (compose syntax). A single container with remark42 can handle multiple sites, no need to run one per site.

as you [setup your websites](https://github.com/umputun/remark42#setup-on-your-website) set appropriate `site_id` values, i.e. `foo` or `bar`

**Reply by ChristopherGS** [reply]
@umputun thank you for clarifying this point. I have a website building platform, so quite a lot of sites (>100) and some of them are running on multiple domains (platform subdomain and custom domain). 

Is the `SITE` parameter purely used to map the `site_id` specified in frontend requests? Or is it used for any kind of CORS check/other logic where the actual url needs to be accurate or 1:1? 

For example if I have:
- `foo.com`
- `foo.myplatform.com`

and then:
- I pass in --site=foo in my server params
- and in the frontend when I make requests from both of the sites I used the `site_id` `foo` 

then is that OK? I wonder if it would mean that comments saved on a particular page at `foo.com/baz` would not be fetched when a different user visited `foo.myplatform.com/baz`?

**Reply by paskal** [reply]
Apologies wrote an answer from the wrong computer initially. Here is it.

> Is the SITE parameter purely used to map the site_id specified in frontend requests? 

Correct, and data is stored on the backend per `site_id`. Users authenticated on one site would be logged in on all others with the same `site_id`.

> Or is it used for any kind of CORS check/other logic where the actual url needs to be accurate or 1:1?

If you have all hosts in `--allowed-hosts`, then they all will be added to `Content-Security-Policy "frame-ancestors ..."`, and you'll have no CORS headers at all in case `--proxy-cors` is set to true.

> I wonder if it would mean that comments saved on a particular page at foo.com/baz would not be fetched when a different user visited foo.myplatform.com/baz?

Comment page resolution works for full URL, so the collision of the same path on two different domains is impossible. While users wouldn't be able to see comments from other sites, technically, they could get these from the API. Still, it's unlikely what they will do unless you create the "last comment widget", which will do "fetch last N comments" request and get these from all pages under the same `site_id` despite the domain name.

**Reply by hao-lee** [reply]
@umputun Actually, `--site=foo,bar` doesn't work because it will treat `foo,bar` as one site.

**Reply by paskal** [reply]
> @umputun Actually, `--site=foo,bar` doesn't work because it will treat `foo,bar` as one site.

Correct, comma works only for environment variables, in CLI options you need to pass it multiple times: '--site=foo --site=bar'.

**Reply by hao-lee** [reply]
@paskal Wow, Got it ! Thanks!

**Comment by umputun** [comment]
note: we have [Discussions](https://github.com/umputun/remark42/discussions) enabled and it is a better place to ask questions.

thx

**Comment by mipimipi** [comment]
Will do. Thx.

**Comment by thunder-spb** [comment]
Hey @umputun , is there any example how to define `ALLOWED_HOSTS`? If I define hostnames comma separated, this does not work. I'm just defining hostnames, e.g `ALLOWED_HOSTS=example1.com,example2.com,example3.com`

**Reply by umputun** [reply]
this is a comma-separated list of fqdns, so your example seems to be fine.

You should see in the log a message generated by this `log.Printf("[INFO] allowed from %+v only", s.AllowedAncestors)` and get a header `	w.Header().Set("Content-Security-Policy", "frame-ancestors "+strings.Join(hosts, " ")+";")` in each response

**Reply by thunder-spb** [reply]
hm, weird, if I set those, iframe could not load... will check logs then... thanks.

**Comment by hao-lee** [comment]
How can I specify different options for two sites? For example, I want to use anonymous login for `foo` while use email login for `bar`.

**Reply by paskal** [reply]
It's not possible at the moment, login settings are instance-wide without ability to set more granular options per-site.

**Reply by hao-lee** [reply]
@paskal Got it. Thanks 😅

**Comment by RJanssens** [comment]
When defining multiple sites, is it possible to set the admin's email address for notifications on a per-site basis? I'd like to use the same back-end for multiple sites, and have a different person responsible for following up on comments for each site.

**Reply by umputun** [reply]
This is possible with a custom AdminStore implemented **outside of the core** Remark42. The core version has `shared` admin type which is the same for all the sites.
